### PR TITLE
Zygote for flow_constraints

### PR DESCRIPTION
**Pull request by michel2323**
The GPU gradient test fails with slightly different values. The difference is too big though to be ignored.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -18,6 +18,7 @@ Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
 SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
 SparseDiffTools = "47a9eef4-7e08-11e9-0b38-333d64bd3804"
 TimerOutputs = "a759f4b9-e2f1-59dc-863e-4aeb61b1ea8f"
+Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"
```
Comment (frapac) [inline Project.toml:21]: Could we add a `compat` entry to Zygote?

File: src/models/polar/constraints.jl [hunk 2]
```diff
@@ -207,6 +207,34 @@ function flow_constraints(polar::PolarForm, cons, buffer)
...
+    to_flow = similar(cons)
+    fr_vmag = buffer.vmag[f]
+    to_vmag = buffer.vmag[t]
+    Δθ = buffer.vang[f] .- buffer.vang[t]
+    cosθ = cos.(Δθ)
+    sinθ = sin.(Δθ)
+    function lumping(vmag, vang)
+        fr_vmag = vmag[f]
+        to_vmag = vmag[t]
+        Δθ =vang[f] .- vang[t]
```
Comment (frapac) [inline src/models/polar/constraints.jl:225]: just for linting :angel: 
could we add a space before `=`?

File: src/models/polar/kernels.jl [hunk 3]
```diff
@@ -414,3 +414,34 @@ end
     slines[ℓ + nlines] = to_flow
 end
 
+function branch_flow_kernel_zygote(
```
Comment (frapac) [inline src/models/polar/kernels.jl:417]: I like this vectorized function. Do you know how many allocations does it lead to on the GPU?

File: test/polar_form.jl [hunk 4]
```diff
@@ -107,6 +107,31 @@ const PS = PowerSystem
...
+                g = VT(undef, m) 
+                ExaPF.flow_constraints(polar, g, adcache)
+                return sum(g)
+            end
+            gradg = ForwardDiff.gradient(lumping,x)
+            ## We pick sum() as the reduction function. This could be a mask function for active set or some log(x) for lumping.
+            result = ExaPF.flow_constraints_grad(polar, g, cache, sum)
+            gradg_zy = vcat(result[1], result[2])
+            # Verify  ForwardDiff and Zygote agree on the gradient
+            @test isapprox(gradg, gradg_zy)
```
Comment (frapac) [inline test/polar_form.jl:134]: Nice that it works already on the CPU!

**Comment by michel2323** [comment]
We have a race condition on the GPU for views:
https://github.com/FluxML/Zygote.jl/blob/2b17256e79b2eca9a6512207284219d279398fc9/src/lib/array.jl#L35
https://github.com/JuliaLang/julia/pull/31407
https://github.com/JuliaGPU/CUDA.jl/issues/89

File: Project.toml (hunk 1 above)
Comment (michel2323) [reply Project.toml:21]: Sure, I haven't used that much before.

File: src/models/polar/constraints.jl (hunk 2 above)
Comment (michel2323) [reply src/models/polar/constraints.jl]: Yea, well this has changed a little now 😆

File: src/models/polar/kernels.jl (hunk 3 above)
Comment (michel2323) [reply src/models/polar/kernels.jl:417]: I will look into this.

File: test/polar_form.jl (hunk 4 above)
Comment (michel2323) [reply test/polar_form.jl:135]: And GPU now :-)

**Comment by michel2323** [comment]
This is still a prototype. `accumulate_view!()` should go into kernels.jl.

File: test/polar_form.jl [hunk 5]
```diff
@@ -107,6 +107,32 @@ const PS = PowerSystem
...
+                g = VT(undef, m) 
+                ExaPF.flow_constraints(polar, g, adcache)
+                return sum(g)
+            end
+            gradg = ForwardDiff.gradient(lumping,x)
+            ## We pick sum() as the reduction function. This could be a mask function for active set or some log(x) for lumping.
+            gradg_zy = ExaPF.flow_constraints_grad(polar, g, cache, sum)
+            # Verify  ForwardDiff and Zygote agree on the gradient
+            
+            # device == CUDADevice() ? (@test_broken isapprox(gradg, gradg_zy)) : (@test isapprox(gradg, gradg_zy))
```
Comment (frapac) [inline test/polar_form.jl:134]: Maybe remove dead line there?

File: src/models/polar/constraints.jl [hunk 6]
```diff
@@ -207,6 +207,50 @@ function flow_constraints(polar::PolarForm, cons, buffer)
     wait(ev)
     return
 end
+
+function flow_constraints_grad(polar::PolarForm, cons, buffer, reduction::Function=sum)
+    function accumulate_view!(x, vx, indices)
```
Comment (frapac) [inline src/models/polar/constraints.jl:212]: Could we add a comment to explain the exact purpose of `accumulate_view!`?

File: src/models/polar/constraints.jl [hunk 7]
```diff
@@ -207,6 +207,50 @@ function flow_constraints(polar::PolarForm, cons, buffer)
...
+        sinθ = sin.(Δθ)
+        cons = branch_flow_kernel_zygote(
+            polar.topology.yff_re, polar.topology.yft_re, polar.topology.ytf_re, polar.topology.ytt_re,
+            polar.topology.yff_im, polar.topology.yft_im, polar.topology.ytf_im, polar.topology.ytt_im,
+            fr_vmag, to_vmag,
+            cosθ, sinθ
+        )
+        return reduction(cons)
+    end
+    grad = Zygote.gradient(lumping, fr_vmag, to_vmag, fr_vang, to_vang)
```
Comment (frapac) [inline src/models/polar/constraints.jl:243]: I have a naive question: what is the type of the output `gradient`? Do we have a tuple with one array by argument?

File: src/models/polar/constraints.jl [hunk 8]
```diff
@@ -207,6 +207,50 @@ function flow_constraints(polar::PolarForm, cons, buffer)
...
+    gvmag = similar(buffer.vmag)
+    gvang = similar(buffer.vang)
```
Comment (frapac) [inline src/models/polar/constraints.jl:245]: This is not a priority for this PR, but I would maybe add a `TODO` to tell that we would need to move `gvmag` and `gvang` in a buffer.

File: test/polar_form.jl [hunk 9]
```diff
@@ -132,7 +132,8 @@ const PS = PowerSystem
             zygradg = ExaPF.flow_constraints_grad(polar, cache, sum)
             # Verify  ForwardDiff and Zygote agree on the gradient
             @test isapprox(adgradg, fdgradg)
-            @test isapprox(adgradg, zygradg)
+            # This breads because of issue #89
```
Comment (frapac) [inline test/polar_form.jl:135]: `This breads and cheese`? You sounds like French :-)

**Review by frapac — APPROVED** [review]

**Comment by frapac** [comment]
I will merge this manually

**Pull request merged**